<a href="https://colab.research.google.com/github/kusuma687/AIAC/blob/main/lab3_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Task1:Try 3 different prompts to generate a factorial function.

prompt1:Write a function in Python that calculates the
factorial of a non-negative integer using recursion.

prompt2:Create a Python function that computes the factorial of a given number using a for loop.

prompt3:Implement a factorial function in Python using the method and lambda function.


In [2]:
def recursive_factorial(n):
    """
    Calculates the factorial of a non-negative integer using recursion.

    Args:
        n: A non-negative integer.

    Returns:
        The factorial of n.
    """
    if n == 0:
        return 1
    else:
        return n * recursive_factorial(n-1)

# Example usage:
num = 5
print(f"The factorial of {num} using recursion is: {recursive_factorial(num)}")

The factorial of 5 using recursion is: 120


Explanation:The code defines a function recursive_factorial(n) that calculates the factorial of a non-negative integer n using recursion.

The base case is when n is 0, in which case it returns 1. This stops the recursion.
The recursive step for n > 0 is to return n multiplied by the result of calling recursive_factorial with n-1. This means the function calls itself with a smaller argument until it reaches the base case.
The example usage shows how to call the function with num = 5 and prints the result. The function will calculate 5 * 4 * 3 * 2 * 1, which is 120.

In [3]:
def for_loop_factorial(n):
    """
    Calculates the factorial of a non-negative integer using a for loop.

    Args:
        n: A non-negative integer.

    Returns:
        The factorial of n.
    """
    if n == 0:
        return 1
    else:
        result = 1
        for i in range(1, n + 1):
            result *= i
        return result

# Example usage:
num = 5
print(f"The factorial of {num} using a for loop is: {for_loop_factorial(num)}")

The factorial of 5 using a for loop is: 120


Explanation:This code defines a Python function called for_loop_factorial that calculates the factorial of a non-negative integer n using a for loop.

def for_loop_factorial(n):: This line defines the function named for_loop_factorial that takes one argument, n.
"""Docstring""": This is a docstring that explains what the function does, its arguments, and what it returns.
if n == 0:: This handles the base case where the input n is 0. The factorial of 0 is 1, so the function returns 1.
else:: If n is not 0, the code enters this block.
result = 1: A variable result is initialized to 1. This variable will store the calculated factorial.
for i in range(1, n + 1):: This starts a for loop that iterates from 1 up to and including n.
result *= i: In each iteration of the loop, the current value of result is multiplied by the loop variable i. This accumulates the product of all numbers from 1 to n.
return result: After the loop finishes, the function returns the final value of result, which is the factorial of n.
num = 5: This line sets a variable num to 5 for demonstration.
print(f"..."): This line calls the for_loop_factorial function with num and prints the result in a formatted string.
In summary, the code iteratively multiplies numbers from 1 up to the given input n to compute the factorial.

In [4]:
import math

factorial_lambda = lambda n: math.factorial(n)

# Example usage:
num = 5
print(f"The factorial of {num} using lambda and math.factorial is: {factorial_lambda(num)}")

The factorial of 5 using lambda and math.factorial is: 120


Explanation:This code uses the math.factorial() function, which is built into Python, to calculate the factorial.

The line factorial_lambda = lambda n: math.factorial(n) creates a small, anonymous function called factorial_lambda. This function takes one input, n, and its job is simply to call math.factorial(n) and return the result.

So, when you call factorial_lambda(5), it's essentially the same as calling math.factorial(5), which gives you the factorial of 5. This is a very quick and direct way to get the factorial because Python's math.factorial is highly optimized.

Task2:Provide a clear example input-output prompt to generate a sorting function.

prompt:input write a python code that generates the sorting function only.output write a python code to generate output of sorting function.

In [10]:
# input-write a python code that generates the sorting function only.
def sort_list(data):
    """
    Sorts a list in ascending order using the built-in sorted function.

    Args:
        data: A list.

    Returns:
        A new list containing the sorted elements.
    """
    return sorted(data)

# output-write a python code to generate output of sorting function.
my_list = [3, 1, 4, 1, 5, 9, 2, 6]
sorted_list = sort_list(my_list)
print(sorted_list)

[1, 1, 2, 3, 4, 5, 6, 9]


Exaplanation:This code defines a function called sort_list that takes a list as input. Inside the function, it uses Python's built-in sorted() function, which is a very efficient way to sort a list. sorted() returns a new sorted list without changing the original one.

Then, the code shows an example: it creates a list my_list, calls sort_list with this list, and stores the result in sorted_list. Finally, it prints sorted_list to show the sorted output.

Task3:Start with the vague prompt “Generate python code to calculate power bill” and
improve it step-by-step.

prompt:Generate python code to calculate power bill” and improve it step-by-step.

In [11]:
def calculate_power_bill_basic(kwh_consumption, rate_per_kwh):
    """
    Calculates a basic power bill based on consumption and a single rate.

    Args:
        kwh_consumption: The total energy consumed in kilowatt-hours (kWh).
        rate_per_kwh: The cost per kilowatt-hour.

    Returns:
        The total power bill amount.
    """
    return kwh_consumption * rate_per_kwh

# Example usage:
consumption = 250  # kWh
rate = 0.15       # $ per kWh
bill_amount = calculate_power_bill_basic(consumption, rate)
print(f"Basic power bill amount: ${bill_amount:.2f}")

Basic power bill amount: $37.50


Expalanation:This code defines a function called calculate_power_bill_basic that takes the amount of energy consumed (kwh_consumption) and the cost per unit of energy (rate_per_kwh) as input.

The function simply multiplies the consumption by the rate to get the total bill amount.

The example part of the code sets a sample consumption and rate, calls the function with these values, and then prints the calculated bill amount.

It's a very simple calculation for a power bill where there's just one price for all the energy used.

Task4:Write structured comments to help AI generate two linked functions(e.g., login_user()
and register_user()).

prompt:write a python code to generate structured comments to help AI generate two linked functions.

In [15]:
# Goal: Generate two linked Python functions: register_user and login_user.

# --- Function 1: register_user ---
# This function should handle new user registration.
# It needs to accept a username and a password as input.
# Before registering, it must check if the username already exists in a storage (e.g., a dictionary or database).
# If the username exists, registration should fail.
# If the username does not exist, the username and a SECURELY HASHED version of the password should be stored.
# The function should return True if registration is successful, False otherwise.

def register_user(username, password):
    # AI: Implement registration logic, including username check and password hashing/storage.
    pass # Placeholder

# --- Function 2: login_user ---
# This function should handle user authentication.
# It needs to accept a username and a password as input.
# It must look up the provided username in the storage where user data is kept.
# If the username is found, it must compare the provided password with the stored HASHED password.
# This comparison should use a secure method (e.g., comparing hashes).
# If the username is not found OR the password does not match the stored hash, login should fail.
# The function should return True if login is successful (username found and password matches), False otherwise.

def login_user(username, password):
    # AI: Implement login logic, including user lookup and secure password verification.
    pass # Placeholder

# AI: After generating the functions, please also provide a simple example demonstrating
# how to use both register_user and login_user, showing both successful and failed cases.

Explanation:This cell contains comments and basic function definitions (def register_user(...) and def login_user(...)). The comments are written in a structured way to tell an AI what these functions should do.

For example, the comments for register_user explain that it should take a username and password, check if the username is already used, and securely store the user's information if the username is new.

Similarly, the comments for login_user explain that it should take a username and password and check if they match the stored information for an existing user.

The pass inside each function is just a placeholder, indicating that the actual code to perform these actions needs to be added. The goal of this cell is to provide clear instructions so that an AI can then write the complete code for these two linked functions.